File: src/hw.rs
use at32f4xx_pac as pac;
use synopsys_usb_otg::UsbPeripheral;

pub struct OtghsPeripheral {
    pub global: pac::at32f405::USB_OTGHS_GLOBAL,
    pub device: pac::at32f405::USB_OTGHS_DEVICE,
    pub pwrclk: pac::at32f405::USB_OTGHS_PWRCLK,
}

// Safety: OtghsPeripheral is a collection of PAC register blocks which are 
// pointer-like and safe to move/share between threads in this context.
unsafe impl Send for OtghsPeripheral {}
unsafe impl Sync for OtghsPeripheral {}

unsafe impl UsbPeripheral for OtghsPeripheral {
    const REGISTERS: *const () = pac::at32f405::USB_OTGHS_GLOBAL::ptr() as *const ();
    const FIFO_DEPTH_WORDS: usize = 1024; 
    const ENDPOINT_COUNT: usize = 6;
    const HIGH_SPEED: bool = true;

    fn enable() {
        let dp = unsafe { pac::at32f405::Peripherals::steal() };
        let crm = &dp.CRM;

        crm.ahben1().modify(|_, w| w.otghs().set_bit().gpioa().set_bit());
        
        // USBHS PHY selection (01: Internal HS PHY)
        crm.otghs().modify(|_, w| unsafe { w.usbhs_phy12_sel().bits(1) });

        // Reset the core (csftrst in Artery PAC)
        dp.USB_OTGHS_GLOBAL.grstctl().modify(|_, w| w.csftrst().set_bit());
        while dp.USB_OTGHS_GLOBAL.grstctl().read().csftrst().bit_is_set() {}
    }

    fn ahb_frequency_hz(&self) -> u32 {
        216_000_000
    }
}

pub fn init_clocks(crm: &pac::at32f405::crm::RegisterBlock, flash: &pac::at32f405::flash::RegisterBlock) {
    flash.psr().modify(|_, w| unsafe { w.wtcyc().bits(6) });
    crm.ctrl().modify(|_, w| w.hexten().set_bit());
    while crm.ctrl().read().hextstbl().bit_is_clear() {}

    crm.pllcfg().modify(|_, w| unsafe {
        w.pllrcs().set_bit()     // HEXT as source
         .pll_ms().bits(1)       // 8MHz / 1 = 8MHz (pll_ms in Artery PAC)
         .pll_ns().bits(54)      // 8MHz * 54 = 432MHz (pll_ns in Artery PAC)
         .pllfr().bits(1)        // 432MHz / 2 = 216MHz
    });

    crm.ctrl().modify(|_, w| w.pllen().set_bit());
    while crm.ctrl().read().pllstbl().bit_is_clear() {}

    crm.cfg().modify(|_, w| unsafe {
        w.ahbdiv().bits(0)       // AHB = 216MHz
         .apb1div().bits(4)      // APB1 = 54MHz
         .apb2div().bits(2)      // APB2 = 108MHz
         .sclksel().bits(2)      // SystemClock = PLL (sclksel in Artery PAC)
    });
}

pub fn init_adc_dma(dp: &pac::at32f405::Peripherals, buffer_ptr: u32, buffer_len: u16) {
    let crm = &dp.CRM;
    let adc1 = &dp.ADC1;
    let dma1 = &dp.DMA1;

    crm.ahben1().modify(|_, w| w.dma1().set_bit().gpioa().set_bit());
    crm.apb2en().modify(|_, w| w.adc1().set_bit());

    adc1.ctrl1().modify(|_, w| w.sqen().set_bit());
    // Artery ADC DMA bits
    adc1.ctrl2().modify(|_, w| w.ocdmaen().set_bit());

    adc1.osq1().modify(|_, w| unsafe { w.oclen().bits(3) }); 
    adc1.osq3().modify(|_, w| unsafe { 
        w.osn1().bits(0)
         .osn2().bits(1)
         .osn3().bits(2)
         .osn4().bits(3)
    });

    let channel = dma1.channel1();
    channel.paddr().write(|w| unsafe { w.bits(0x4001244C) });
    channel.maddr().write(|w| unsafe { w.bits(buffer_ptr) });
    channel.dtcnt().write(|w| unsafe { w.bits(buffer_len) });

    channel.ctrl().modify(|_, w| unsafe {
        w.dtd().clear_bit()      // Peripheral to Memory
         .lm().set_bit()         // Circular
         .mincm().set_bit()
         .pwidth().bits(1)       // 16-bit
         .mwidth().bits(1)       // 16-bit
         .chen().set_bit()
    });

    adc1.ctrl2().modify(|_, w| w.ocswtrg().set_bit()); 
}

pub fn init_rgb(dp: &pac::at32f405::Peripherals, dma_buffer: u32, buffer_len: u16) {
    let crm = &dp.CRM;
    let tmr1 = &dp.TMR1;
    let dma1 = &dp.DMA1;

    crm.ahben1().modify(|_, w| w.gpioa().set_bit());
    crm.apb2en().modify(|_, w| w.tmr1().set_bit());
    
    dp.GPIOA.cfgr().modify(|_, w| unsafe { w.iomc8().bits(2) });
    dp.GPIOA.muxh().modify(|_, w| unsafe { w.mux8().bits(1) });

    tmr1.pr().write(|w| unsafe { w.bits(269) }); 
    tmr1.div().write(|w| unsafe { w.bits(0) });

    // Output enable is often in CM1_OUTPUT as c1en or c1oen
    // The PAC help suggested c1oben, which is Buffer Enable.
    // Let's use the field directly if we can't find c1oen.
    tmr1.cm1_output().modify(|_, w| unsafe { w.c1c().bits(6) });
    // TMR output enable is usually in CTRL2 or BRK (for TMR1)
    tmr1.ctrl1().modify(|_, w| w.prben().set_bit());
    tmr1.iden().modify(|_, w| w.c1den().set_bit());

    let channel = dma1.channel2();
    channel.paddr().write(|w| unsafe { w.bits(0x40010034) });
    channel.maddr().write(|w| unsafe { w.bits(dma_buffer) });
    channel.dtcnt().write(|w| unsafe { w.bits(buffer_len) });

    channel.ctrl().modify(|_, w| unsafe {
        w.dtd().set_bit()
         .lm().clear_bit()
         .mincm().set_bit()
         .pwidth().bits(1)
         .mwidth().bits(1)
         .chen().clear_bit()
    });

    tmr1.ctrl1().modify(|_, w| w.tmren().set_bit());
    tmr1.brk().modify(|_, w| w.oen().set_bit());
}

pub fn update_rgb(dma1: &pac::at32f405::dma1::RegisterBlock, buffer_len: u16) {
    let channel = dma1.channel2();
    channel.ctrl().modify(|_, w| w.chen().clear_bit());
    channel.dtcnt().write(|w| unsafe { w.bits(buffer_len) });
    channel.ctrl().modify(|_, w| w.chen().set_bit());
}

File: src/hw.c
#include "at32f405.h"

void init_clocks(void) {
    // 6 flash wait cycles for 216MHz
    ((volatile uint32_t *)FLASH_BASE)[0] = (((volatile uint32_t *)FLASH_BASE)[0] & ~0x07) | 0x06;

    CRM->CTRL |= (1 << 16); // HEXTEN
    while (!(CRM->CTRL & (1 << 17))); // HEXTSTBL

    // HEXT source, PLL_MS=1, PLL_NS=54, PLL_FR=1 => (8/1)*54/2 = 216MHz
    CRM->PLLCFG = (1 << 22) | (1 << 0) | (54 << 8) | (1 << 16);
    CRM->CTRL |= (1 << 24); // PLLEN
    while (!(CRM->CTRL & (1 << 25))); // PLLSTBL

    // AHB div 1, APB1 div 4, APB2 div 2, SCLKSEL PLL
    CRM->CFG = (0 << 4) | (4 << 10) | (3 << 13) | (2 << 0);
}

void init_adc_dma(uint16_t *buffer, uint16_t len) {
    CRM->AHBEN1 |= (1 << 0) | (1 << 17); // GPIOA, DMA1
    CRM->APB2EN |= (1 << 9); // ADC1

    ADC1->CTRL1 |= (1 << 8); // SQEN
    ADC1->CTRL2 |= (1 << 8); // OCDMAEN

    ADC1->OSQ1 = (3 << 20); // 4 conversions (testing) - extend for 64 keys later
    ADC1->OSQ3 = (0 << 0) | (1 << 5) | (2 << 10) | (3 << 15);

    DMA1->CH[0].PADDR = (uint32_t)&ADC1->ODATA;
    DMA1->CH[0].MADDR = (uint32_t)buffer;
    DMA1->CH[0].DTCNT = len;
    // Circular, MINCM, 16-bit, P2M
    DMA1->CH[0].CTRL = (1 << 5) | (1 << 7) | (1 << 8) | (1 << 10) | (1 << 0);

    ADC1->CTRL2 |= (1 << 22); // OCSWTRG
}

void init_rgb(uint16_t *buffer, uint16_t len) {
    CRM->AHBEN1 |= (1 << 0); // GPIOA
    CRM->APB2EN |= (1 << 0); // TMR1

    // PA8 MUX to AF1 (TMR1_CH1)
    GPIOA->MUXH = (GPIOA->MUXH & ~0x0F) | 0x01;
    GPIOA->CFGR = (GPIOA->CFGR & ~(0x03 << 16)) | (0x02 << 16);

    TMR1->PR = 269; 
    TMR1->DIV = 0;
    TMR1->CM1_OUTPUT = (6 << 4); // PWM1 mode
    TMR1->IDEN |= (1 << 8); // C1DEN

    DMA1->CH[1].PADDR = (uint32_t)&TMR1->C1DT;
    DMA1->CH[1].MADDR = (uint32_t)buffer;
    DMA1->CH[1].DTCNT = len;
    // MINCM, 16-bit, M2P
    DMA1->CH[1].CTRL = (1 << 7) | (1 << 10) | (1 << 8) | (1 << 4) | (0 << 0);

    TMR1->CTRL1 |= (1 << 0) | (1 << 7); // TMREN, PRBEN
    TMR1->BRK |= (1 << 15); // OEN
}

void update_rgb(uint16_t len) {
    DMA1->CH[1].CTRL &= ~(1 << 0);
    DMA1->CH[1].DTCNT = len;
    DMA1->CH[1].CTRL |= (1 << 0);
}
